### PR TITLE
feature(wgt): enable DI using torch-rpc to support GPU-p2p and RDMA-rpc

### PR DESCRIPTION
**Pull request by SolenoidWGT**
Commit:
1. Add torchrpc message queue.

2. Implement buffer based on CUDA-shared-tensor to optimize the data path of torchrpc.

3. Add 'bypass_eventloop' arg in Task() and Parallel().

4. Add thread lock in distributer.py to prevent sender and receiver competition.

5. Add message queue perf test for torchrpc, nccl, nng, shm

6. Add comm_perf_helper.py to make program timing more convenient.

7. Modified the subscribe() of class MQ, adding 'fn' parameter and 'is_once' parameter.

8. Add new DummyLock and ConditionLock type in lock_helper.py

9. Add message queues perf test.

10. Introduced a new self-hosted runner to execute cuda, multiprocess, torchrpc related tests.

## Description
DI-engine integrates torch.distributed.rpc module.
   1. CPU-P2P-RDMA: In IB network environment, support RDMA CPU-P2P transmission
   2. GPU-P2P-RDMA: supports GPU p2p communication

### cli-ditask introduces new command line arguments

1. `--mq_type`: Introduced `torchrpc:cuda` and `torchrpc:cpu` options

   1. `torchrpc:cuda`: Use torchrpc for communication, and allow setting device_map, can use GPU direct RDMA.
   2. ` torchrpc:cpu:` Use torchrpc for communication, but device_map is not allowed to be set. All data on the GPU side will be copied to the CPU side for transmission.

2. `--init-method`: Initialization entry for init_rpc (required if --mq_type is torchrpc)

3. `--local-cuda-devices`: Set the rank range of local GPUs that can be used (optional, default is all visible devices)

4. `--cuda-device-map`: Used to set device_map, the format is as follows:

   ```
   Format:
   <Peer node id>_<Local GPU rank>_<Peer GPU rank>,[...]
   example:
   --cuda-device-map=1_0_1,2_0_2,0_0_0
   
   pytorch api:
   options.set_device_map("worker1", {1: 2})
   ```

   (Optional, the default is to map all visible GPU to the GPU-0 of the peer)

### Dynamic GPU communication groups

We create devices mappings between all possible devices in advance. This mapping is all-2-all, which can cover all communication situations. The purpose is to avoid errors caused by incomplete devicemap coverage. Setting redundant mappings will not have any side effects. The mappings are used to check the validity of the device during transport. Only after a new process joins the communication group will it try to create a channel based on these maps.

```
Node_0 device_maps:
("Node_1", {0: 0}), ("Node_2", {0: 0}), ...., ("Node_99", {0: 0})

Node_1 device_maps:
("Node_0", {0: 0}), ("Node_2", {0: 0}), ...., ("Node_99", {0: 0})
```

At the same time, we still expose the `--cuda-device-map` interface, which is used to allow users to configure the topology between devices, torchrpc will follow user input.
## TODO
Load balancing capability: in a time-heterogeneous RL task environment, each worker can run at full capacity.

## Check List

- [x] merge the latest version source branch/repo, and resolve all the conflicts
- [x] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: ding/data/tests/test_shm_buffer.py [hunk 1]
```diff
@@ -1,20 +1,90 @@
...
+@pytest.mark.cudatest
+# @pytest.mark.multiprocesstest
+def test_cuda_shm():
+    if torch.cuda.is_available() and torch.cuda.device_count() >= 2:
+        import torch.multiprocessing as mp
+        ctx = mp.get_context('spawn')
+
+        event_run = ctx.Event()
+        shm_buf_np = ShmBufferCuda(np.dtype(np.float32), shape=(1024, 1024), copy_on_get=True)
+        shm_buf_torch = ShmBufferCuda(torch.float32, shape=(1024, 1024), copy_on_get=True)
```
Comment (PaParaZz1) [inline ding/data/tests/test_shm_buffer.py:59]: we should add another unittest for the case `copy_on_get=False` to validate it.

File: ding/data/shm_buffer.py [hunk 2]
```diff
@@ -78,6 +111,94 @@ def get(self) -> np.ndarray:
...
+        with torch.no_grad():
+            self.buffer.copy_(tensor.view(tensor.numel()))
+
+    def get(self) -> Union[np.ndarray, torch.Tensor]:
+        with torch.no_grad():
+            if self.ctype is np.ndarray:
+                # Because ShmBufferCuda use CUDA memory exchanging data between processes.
+                # So copy_on_get is necessary for numpy arrays.
+                re = self.buffer.cpu()
+                re = re.detach().view(self.shape).numpy()
```
Comment (PaParaZz1) [inline ding/data/shm_buffer.py:189]: why only detach rather than clone().detach here

File: ding/entry/cli_ditask.py [hunk 3]
```diff
@@ -145,6 +172,18 @@ def _cli_ditask(
     if node_ids and not isinstance(node_ids, int):
         node_ids = node_ids.split(",")
         node_ids = list(map(lambda i: int(i), node_ids))
+    use_cuda = False
+    if mq_type == "torchrpc:cuda" or mq_type == "torchrpc:cpu":
+        mq_type, use_cuda = mq_type.split(":")
+        if use_cuda == "cuda":
+            use_cuda = True
```
Comment (PaParaZz1) [inline ding/entry/cli_ditask.py:179]: just name as `cuda`

File: ding/envs/env_manager/subprocess_env_manager.py [hunk 4]
```diff
@@ -97,6 +100,7 @@ def __init__(
         """
         super().__init__(env_fn, cfg)
         self._shared_memory = self._cfg.shared_memory
+        self._cuda_shared_memory = self._cfg.cuda_shared_memory if self._shared_memory else False
```
Comment (PaParaZz1) [inline ding/envs/env_manager/subprocess_env_manager.py:103]: maybe we should raise warning if `self._shared_memory=False` but `self._cfg.cuda_shared_memory=True`

File: ding/framework/message_queue/README.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+### 1. container environment
+
+Tensorpipe is not container aware. Processes can find themselves on the same physical machine through `/proc/sys/kernel/random/boot_id` ,but because in separated pod/container, they cannot use means of communication such as CUDA ipc. When tensorpipe finds that these communication methods cannot be used, it will report an error and exit. 
+
+### 2. RDMA and fork subprocess
+
+Tensorpipe does not consider the case of calling [fork(2)](https://man7.org/linux/man-pages/man2/fork.2.html) when using RDMA. If the corresponding initialization measures are not performed when using RDMA, using fork will cause serious problems, refer to [here](https://www.rdmamojo.com/2012/05/24/ibv_fork_init/). Therefore, if you start ditask in the IB/RoCE network environment, please specify the environment variables `IBV_FORK_SAFE=1` and `RDMAV_FORK_SAFE=1` , so that ibverbs will automatically initialize fork support.
```
Comment (PaParaZz1) [inline ding/framework/message_queue/README.md:13]: add a basic performance comparison table

File: ding/framework/message_queue/perfs/perf_shm.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    # Step2: gpu shared tensor -> gpu tensor
+    tensor = buffers[payload.data["idx"]].get()
+    assert tensor.device == torch.device('cuda:0')
+    # Step3: gpu tensor(cuda:0) -> gpu tensor(cuda:1)
+    tensor = tensor.to(1)
+    torch.cuda.synchronize(1)
+    assert tensor.device == torch.device('cuda:1')
+
+
+class Recvier:
```
Comment (PaParaZz1) [inline ding/framework/message_queue/perfs/perf_shm.py:40]: Receiver

File: dizoo/atari/example/atari_dqn_dist_rdma.py [hunk 7]
```diff
@@ -26,46 +28,53 @@ def main():
...
-            task.use(model_exchanger(model, is_learner=False))
+            task.use(ContextExchanger(skip_n_iter=1))
+            task.use(ModelExchanger(model))
             task.use(eps_greedy_handler(cfg))
             task.use(StepCollector(cfg, policy.collect_mode, collector_env))
             task.use(termination_checker(max_env_step=int(1e7)))
         else:
             raise KeyError("invalid router labels: {}".format(task.router.labels))
 
-        task.run()
```
Comment (PaParaZz1) [inline dizoo/atari/example/atari_dqn_dist_rdma.py:68]: why remove this

File: ding/utils/lock_helper.py [hunk 8]
```diff
@@ -118,3 +136,24 @@ def get_file_lock(name: str, op: str) -> None:
...
                 pass
         return FcntlContext(lock_name)
+
+
+def synchronized(func):
+    """
+    Overview:
+        thread lock decorator.
+    Arguments:
+        - func ([type]): A function that needs to be protected by a lock.
```
Comment (PaParaZz1) [inline ding/utils/lock_helper.py:146]: Callable

File: ding/utils/comm_perf_helper.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+
+
+def _store_timer_result(func_name: str, avg_tt: float):
+    if func_name not in OUTPUT_DICT.keys():
+        OUTPUT_DICT[func_name] = str(round(avg_tt, 4)) + ","
+    else:
+        OUTPUT_DICT[func_name] = OUTPUT_DICT[func_name] + str(round(avg_tt, 4)) + ","
+
+
+def print_timer_result_csv():
```
Comment (PaParaZz1) [inline ding/utils/comm_perf_helper.py:40]: maybe you can use `pretty_print` function in `ding.utils`

File: ding/framework/task.py [hunk 10]
```diff
@@ -424,51 +430,82 @@ def async_executor(self, fn: Callable, *args, **kwargs) -> None:
...
             - only_remote (:obj:`bool`): Only broadcast the event to the connected nodes, default is False.
             - only_local (:obj:`bool`): Only emit local event, default is False.
+            - bypass_eventloop (:obj:`bool`): Whether to select to bypass eventloop of Task() and Parallel(),
+                this parameter can only be True when torchrpc is used as the communication backend. If use torchrpc,
+                the invoked of the callback is triggered by the torchrpc's backend thread.
             - args (:obj:`any`): Rest arguments for listeners.
         """
         # Check if need to broadcast event to connected nodes, default is True
         assert self._running, "Please make sure the task is running before calling the this method, see the task.start"
-        if only_local:
```
Comment (PaParaZz1) [inline ding/framework/task.py:439]: why remove this

File: ding/framework/middleware/functional/trainer.py [hunk 11]
```diff
@@ -71,8 +71,8 @@ def _train(ctx: Union["OnlineRLContext", "OfflineRLContext"]):
 
         if ctx.train_data is None:  # no enough data from data fetcher
             return
-        data = ctx.train_data.to(policy._device)
-        train_output = policy.forward(data)
+        # data = ctx.train_data.to(policy._device)
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/trainer.py:74]: why comment this

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/framework/middleware/functional/collector.py [hunk 12]
```diff
@@ -113,16 +119,30 @@ def _rollout(ctx: "OnlineRLContext"):
...
             transition = policy.process_transition(ctx.obs[i], ctx.inference_output[i], timestep)
             transition = ttorch.as_tensor(transition)  # TBD
             transition.collect_train_iter = ttorch.as_tensor([ctx.train_iter])
             transition.env_data_id = ttorch.as_tensor([env_episode_id[timestep.env_id]])
+
+            # torchrpc currently uses "cuda:0" as the transmission device by default,
+            # so all data on the cpu side is copied to "cuda:0" here. In fact this
+            # copy is unnecessary, because torchrpc can support both cpu side and gpu
+            # side data to communicate using RDMA, but mixing the two transfer types
+            # will cause a bug, see issue:
```
Comment (PaParaZz1) [inline ding/framework/middleware/functional/collector.py:140]: where is issue

File: .coveragerc [hunk 13]
```diff
@@ -1,4 +1,5 @@
 [run]
+concurrency = multiprocessing,thread
```
[the remark this answers is not in the document]
Comment (SolenoidWGT) [reply .coveragerc:2]: Add concurrency = multiprocessing, so that codecov can count the coverage of subprocesses, and the default concurrency is set to threading. However，there are some things need to pay attention in using, refer to: https://pytest-cov.readthedocs.io/en/latest/subprocess-support.html

File: ding/data/shm_buffer.py (hunk 2 above)
Comment (SolenoidWGT) [reply ding/data/shm_buffer.py:189]: Because self.buffer is located in the GPU, we have already copied it to CPU and there is no need to clone it again.
